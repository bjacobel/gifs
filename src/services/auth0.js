import Auth0Lock from 'auth0-lock';
import { REDIRECT_URL } from '../constants/auth0';

export const isAuthed = (auth) => {
  const { idToken, idTokenExpiry } = auth;
  return idToken && idTokenExpiry && ((new Date()).getTime() / 1000) < idTokenExpiry;
};

export default class Auth0Service {
  constructor(clientId, domain, callback) {
    // Configure Auth0
    this.lock = new Auth0Lock(clientId, domain, {
      theme: {
        logo: 'https://gifs.bjacobel.com/partyparrot.gif',
        primaryColor: '#4376fb',
      },
      languageDictionary: {
        title: 'Gifs',
      },
      auth: {
        redirectUrl: REDIRECT_URL,
        responseType: 'token',
      },
    });
    // Add callback for lock `authenticated` event
    this.lock.on('authenticated', callback);
    // binds login functions to keep this context
    this.login = this.login.bind(this);
  }

  login() {
    // Call the show method to display the widget.
    this.lock.show();
  }
}
