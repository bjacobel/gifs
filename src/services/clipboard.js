// @flow

export const copy = () => {  // eslint-disable-line import/prefer-default-export
  const selection = document.getSelection();

  if (selection) {
    selection.removeAllRanges();
  }

  const copyInput = document.querySelector('.gif-fullname-hidden');

  if (copyInput) {
    copyInput.focus();
    copyInput.setSelectionRange(0, copyInput.value.length);
  }

  if (document.queryCommandSupported('copy') && document.execCommand('copy')) {
    return true;
  } else {
    throw new Error('Copying not supported in your browser');
  }
};
