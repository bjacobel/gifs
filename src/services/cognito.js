import AWS from 'aws-sdk/global';

import {
  AUTHED_ROLE_ARN,
  COGNITO_POOL,
  REGION,
  UNAUTHED_ROLE_ARN,
} from '../constants/aws';
import AuthService from '../services/auth0';
import {
  CLIENT_ID,
  DOMAIN,
} from '../constants/auth0';

export const obtainAuthRole = (token) => {
  return new Promise((resolve, reject) => {
    AWS.config.region = REGION;
    AWS.config.credentials = new AWS.CognitoIdentityCredentials({
      IdentityPoolId: COGNITO_POOL,
    });

    Object.assign(AWS.config.credentials.params, {
      RoleArn: AUTHED_ROLE_ARN,
      Logins: {
        [DOMAIN]: token,
      },
    });

    return AWS.config.credentials.refresh((err) => {
      if (err) {
        reject(err);
      } else {
        resolve(Object.assign({}, AWS.config.credentials.webIdentityCredentials, { isAuthed: true }));
      }
    });
  });
};

export const obtainUnauthedRole = () => {
  return new Promise((resolve, reject) => {
    AWS.config.region = REGION;
    AWS.config.credentials = new AWS.CognitoIdentityCredentials({
      RoleArn: UNAUTHED_ROLE_ARN,
      IdentityPoolId: COGNITO_POOL,
    });

    return AWS.config.credentials.refresh((err) => {
      if (err) {
        reject(err);
      } else {
        resolve(Object.assign({}, AWS.config.credentials.webIdentityCredentials, { isAuthed: false }));
      }
    });
  });
};

export const obtainCurrentRole = () => {
  const auth = new AuthService(CLIENT_ID, DOMAIN);

  if (auth.loggedIn()) {
    return obtainAuthRole(auth.getToken());
  } else {
    // We're not in a position where we can auth, return the unauthed role
    return obtainUnauthedRole();
  }
};
