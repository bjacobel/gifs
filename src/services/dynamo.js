import AWS from 'aws-sdk-umd';
import shortid from 'shortid';

import DYNAMO_TABLE from '../constants/aws';

export const getAllTags = () => {
  return new Promise((resolve, reject) => {
    const dynamo = new AWS.DynamoDB();

    dynamo.scan({
      TableName: DYNAMO_TABLE,
      ProjectionExpression: 'gif_id,tag,id'
    }, (err, data) => {
      if (err) {
        reject(err);
      } else {
        resolve(data);
      }
    });
  });
};

export const addTag = (tag, id) => {
  return new Promise((resolve, reject) => {
    const dynamo = new AWS.DynamoDB({
      credentials: AWS.config.credentials
    });

    const uuid = shortid.generate();

    dynamo.putItem({
      TableName: DYNAMO_TABLE,
      Item: {
        id: { S: uuid },
        gif_id: { S: id },
        tag: { S: tag }
      }
    }, (err) => {
      if (err) {
        reject(err);
      } else {
        // Data returned by dynamo.putItem is null, so just make it up
        resolve({
          id: uuid,
          gif_id: id,
          tag
        });
      }
    });
  });
};

export const deleteTag = (id) => {
  return new Promise((resolve, reject) => {
    const dynamo = new AWS.DynamoDB({
      credentials: AWS.config.credentials
    });
    dynamo.deleteItem({
      TableName: DYNAMO_TABLE,
      Key: {
        id: { S: id }
      }
    }, (err) => {
      if (err) {
        reject(err);
      } else {
        resolve(id);
      }
    });
  });
};
