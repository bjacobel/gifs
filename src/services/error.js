export const redirectToError = (err) => {  // eslint-disable-line import/prefer-default-export
  window.location = `/error#${err.message}`;
};
