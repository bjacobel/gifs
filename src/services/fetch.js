// This approach makes fetch much easier to mock in tests
import 'whatwg-fetch';

export default window.fetch;
