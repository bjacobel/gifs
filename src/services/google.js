import queryString from 'query-string';
import { USER_EMAIL } from '../constants';
import {
  OAUTH_ENDPOINT,
  CLIENT_ID,
  SCOPES
} from '../constants/google';

export const requestAccessToken = () => {
  return new Promise((resolve) => {
    const authURL = `${OAUTH_ENDPOINT}?${queryString.stringify({
      response_type: 'id_token',
      nonce: 'nonce',
      client_id: CLIENT_ID,
      redirect_uri: `${window.location.origin}/googleAuth`,
      scope: SCOPES,
      prompt: 'select_account',
      login_hint: USER_EMAIL
    })}`;
    resolve(window.location.assign(authURL));
  });
};

export const parseGooglePostback = (callback) => {
  if (window.location.pathname === '/googleAuth') {
    const authInfo = queryString.parse(window.location.hash);
    window.history.replaceState({}, null, '/');
    callback(authInfo);
  }
};
