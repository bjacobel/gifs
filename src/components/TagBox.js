import React, { Component } from 'react';
import { connect } from 'react-redux';

import Tag from './Tag';
import { getTagsAsync } from '../actions/tags';
import { ROOT_URL } from '../constants';
import { copy } from '../services/clipboard';

function mapStateToProps(state) {
  return {
    activeGif: state.activeGif,
    gifs: state.gifs,
    tags: state.tags,
  };
}

const mapDispatchToProps = {
  getTagsAsync,
};

class TagBox extends Component {
  componentWillMount() {
    this.props.getTagsAsync();
  }

  render() {
    const { gifs, tags } = this.props;
    const activeGifId = this.props.activeGif;
    const activeGif = gifs.find(gif => gif.id === activeGifId);

    let tagBoxContents;

    if (!activeGifId) {
      const tagValues = Object.values(tags);
      const flatTagList = [].concat(...tagValues);

      tagBoxContents = (
        <div>
          <p><i>{ gifs.length } gifs, { flatTagList.length } tags</i></p>
        </div>
      );
    } else {
      let tagList = null;

      if (tags[activeGifId]) {
        tagList = tags[activeGifId].map((tag) => {
          return (
            <Tag key={ tag.id } id={ tag.id } content={ tag.text } />
          );
        });
      }

      tagBoxContents = (
        <div>
          <input
            type="text"
            className="gif-fullname-hidden"
            value={ ROOT_URL + activeGif.src }
            readOnly
          />
          <p className="gif-name copy-icon" onMouseUp={ () => copy() }>
            <span className="sliding-ul-container">
              <span className="sliding-ul">
                <span className="lightgrey">{ ROOT_URL }</span>
                { activeGif.src }
              </span>
            </span>
          </p>
          <p className="tag-list">
            { tagList }
            <Tag className="adder" meta="add-tag" />
          </p>
        </div>
      );
    }

    return (
      <div className="tag-box">
        { tagBoxContents }
      </div>
    );
  }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(TagBox);
