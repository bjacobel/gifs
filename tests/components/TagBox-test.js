import React from 'react';
import { shallow } from 'enzyme';

import TagBox from '../../src/components/TagBox';
import { copy } from '../../src/services/clipboard';

jest.mock('../../src/services/clipboard');

describe('TagBox component', () => {
  const tags = {
    1: [
      { id: '1', text: 'first tag for 1' },
      { id: '1', text: '2nd tag for 1' },
      { id: '1', text: '3rd tag for 1' },
    ],
  };

  const gifs = [
    { id: '1', src: '1.com' },
    { id: '2', src: '2.com' },
  ];

  it("matches snapshot when the activeGif doesn't have tags", () => {
    expect(
      shallow(
        <TagBox.WrappedComponent
          tags={ tags }
          gifs={ gifs }
          activeGif="2"
          getTagsAsync={ jest.fn() }
        />,
      ),
    ).toMatchSnapshot();
  });

  it('matches snapshot when the activeGif has tags', () => {
    expect(
      shallow(
        <TagBox.WrappedComponent
          tags={ tags }
          gifs={ gifs }
          activeGif="1"
          getTagsAsync={ jest.fn() }
        />,
      ),
    ).toMatchSnapshot();
  });

  it('matches snapshot when no activeGif is set', () => {
    expect(
      shallow(
        <TagBox.WrappedComponent
          tags={ tags }
          gifs={ gifs }
          getTagsAsync={ jest.fn() }
        />,
      ),
    ).toMatchSnapshot();
  });

  it('calls the copy service when you click the name of a gif', () => {
    shallow(
      <TagBox.WrappedComponent
        tags={ tags }
        gifs={ gifs }
        getTagsAsync={ jest.fn() }
        activeGif="1"
      />,
    ).find('.copy-icon').simulate('mouseUp');

    expect(copy).toBeCalled();
  });
});
