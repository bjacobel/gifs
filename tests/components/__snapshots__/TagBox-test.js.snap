exports[`TagBox component matches snapshot when no activeGif is set 1`] = `
<div
  className="tag-box">
  <div>
    <p>
      <i>
        2
         gifs, 
        3
         tags
      </i>
    </p>
  </div>
</div>
`;

exports[`TagBox component matches snapshot when the activeGif doesn't have tags 1`] = `
<div
  className="tag-box">
  <div>
    <input
      className="gif-fullname-hidden"
      readOnly={true}
      type="text"
      value="https://gifs.bjacobel.com/2.com" />
    <p
      className="gif-name copy-icon"
      onMouseUp={[Function]}>
      <span
        className="sliding-ul-container">
        <span
          className="sliding-ul">
          <span
            className="lightgrey">
            https://gifs.bjacobel.com/
          </span>
          2.com
        </span>
      </span>
    </p>
    <p
      className="tag-list">
      <Connect(Tag)
        className="adder"
        meta="add-tag" />
    </p>
  </div>
</div>
`;

exports[`TagBox component matches snapshot when the activeGif has tags 1`] = `
<div
  className="tag-box">
  <div>
    <input
      className="gif-fullname-hidden"
      readOnly={true}
      type="text"
      value="https://gifs.bjacobel.com/1.com" />
    <p
      className="gif-name copy-icon"
      onMouseUp={[Function]}>
      <span
        className="sliding-ul-container">
        <span
          className="sliding-ul">
          <span
            className="lightgrey">
            https://gifs.bjacobel.com/
          </span>
          1.com
        </span>
      </span>
    </p>
    <p
      className="tag-list">
      <Connect(Tag)
        content="first tag for 1"
        id="1" />
      <Connect(Tag)
        content="2nd tag for 1"
        id="1" />
      <Connect(Tag)
        content="3rd tag for 1"
        id="1" />
      <Connect(Tag)
        className="adder"
        meta="add-tag" />
    </p>
  </div>
</div>
`;
